src/shadowbox/server/server_access_key.ts

```diff
@@ -90,73 +123,162 @@ export class ServerAccessKeyRepository implements AccessKeyRepository {
+    const accessKey = this.getAccessKey(id);
+    if (!accessKey) {
+      return false;
+    }
+    const quotaStausChanged = accessKey.isOverQuota;
+    accessKey.quota = undefined;
+    accessKey.isOverQuota = false;
+    try {
+      this.saveAccessKeys();
+      if (quotaStausChanged) {
+        await this.updateServer();
+      }
     } catch (error) {
       return false;
     }
     return true;
   }
 
   getMetricsId(id: AccessKeyId): AccessKeyMetricsId|undefined {
-    const accessKeyJson = this.getAccessKey(id);
-    return accessKeyJson ? accessKeyJson.metricsId : undefined;
+    const accessKey = this.getAccessKey(id);
+    return accessKey ? accessKey.metricsId : undefined;
+  }
+
+  // Compares access key usage with collected metrics, marking them as under or over quota.
+  async enforceAccessKeyQuotas() {
+    let quotaStatusChanged = false;
+    for (const accessKey of this.accessKeys) {
+      quotaStatusChanged = quotaStatusChanged || await this.updateAccessKeyQuotaStaus(accessKey);
+    }
+    if (quotaStatusChanged) {
+      this.updateServer();
+    }
+  }
+
+  // Updates `accessKey` quota status by comparing its usage with collected metrics. Returns whether
+  // the quota status changed.
+  private async updateAccessKeyQuotaStaus(accessKey: AccessKey): Promise<boolean> {
+    if (!accessKey.quota) {
+      return false;  // Don't query the usage of access keys without quota.
+    }
+    const usage =
+        await this.metrics.getOutboundByteTransfer(accessKey.id, accessKey.quota.windowHours);
+    const usageBytes = usage.bytesTransferredByUserId[accessKey.id];
+    const isOverQuota = usageBytes > accessKey.quota.quotaBytes;
+    logging.debug(`Enforcing quota for access key ${accessKey.id}. Quota: ${
+        JSON.stringify(accessKey.quota)}, usage: ${usageBytes}B, isOverQuota: ${isOverQuota}`);
+    const quotaStatusChanged = isOverQuota !== accessKey.isOverQuota;
+    accessKey.isOverQuota = isOverQuota;
+    return quotaStatusChanged;
   }
 
   private updateServer(): Promise<void> {
-    return this.shadowsocksServer.update(this.keyConfig.data().accessKeys.map((e) => {
-      return {id: e.id, port: e.port, cipher: e.encryptionMethod, secret: e.password};
-    }));
+    const serverAccessKeys = this.accessKeys.filter(e => !e.isOverQuota).map(e => {
+      return {
+        id: e.id,
+        port: e.proxyParams.portNumber,
+        cipher: e.proxyParams.encryptionMethod,
+        secret: e.proxyParams.password
+      };
+    });
+    return this.shadowsocksServer.update(serverAccessKeys);
   }
 
-  private getAccessKey(id: AccessKeyId): AccessKeyConfig {
-    for (const accessKeyJson of this.keyConfig.data().accessKeys) {
-      if (accessKeyJson.id === id) {
-        return accessKeyJson;
+  private loadAccessKeys(): AccessKey[] {
+    return this.keyConfig.data().accessKeys.map(e => makeAccessKey(this.proxyHostname, e));
+  }
+
+  private saveAccessKeys() {
```

Comment:
maybe (de)?serializeAccessKeys() is more descriptive?

Reply:
I wanted the name to imply a write/read operation, thus I chose 'save'/'load'. The (de)serialization seems more of an implementation detail.